src/TreeSitter/Node.hs

```diff
@@ -15,78 +20,111 @@ data Node = Node
   , nodeSymbol :: !Word16
   , nodeStartPoint :: !TSPoint
   , nodeEndPoint :: !TSPoint
-  , nodeStartByte :: !Int32
-  , nodeEndByte :: !Int32
-  , nodeNamedChildCount :: !Int32
-  , nodeChildCount :: !Int32
+  , nodeStartByte :: !Word32
+  , nodeEndByte :: !Word32
+  , nodeChildCount :: !Word32
   }
-  deriving (Show, Eq, Generic, CStorable)
+  deriving (Show, Eq, Generic)
 
-data TSPoint = TSPoint { pointRow :: !Int32, pointColumn :: !Int32 }
-  deriving (Show, Eq, Generic, CStorable)
+data TSPoint = TSPoint { pointRow :: !Word32, pointColumn :: !Word32 }
+  deriving (Show, Eq, Generic)
 
+data TSNode = TSNode !(Ptr ()) !Word32 !Word32
+  deriving (Show, Eq, Generic)
 
-data TSNode = TSNode !(Ptr ()) !Int32 !Int32 !Int32
-  deriving (Show, Eq, Generic, CStorable)
 
-peekAdvance :: forall a b. Storable a => Ptr a -> IO (a, Ptr b)
-peekAdvance ptr = do
-  let aligned = alignPtr ptr (alignment (undefined :: a))
+-- | 'Struct' is a strict 'Monad' with automatic alignment & advancing, & inferred type.
+newtype Struct a = Struct { runStruct :: forall b . Ptr b -> IO (a, Ptr a) }
```

Comment:
I briefly experimented with a deep embedding of this DSL with the goal of being able to capture the `alignment` and `sizeOf` operations as well, but it ended up being kind of clumsy to express since peeking had to provide a value before it could evaluate the rest of it. Representing this DSL with an `Applicative` instance but not `Monad` would do nicely, but I was getting off track so I just left it as a shallow embedding.